### PR TITLE
Rewrite typing tests as expect tests

### PR DESCRIPTION
**Pull request by ghost**
Reusing the idea of [ppx_expect](https://github.com/janestreet/ppx_expect), I added an support in the toplevel for checking the expected result of the evaluation of an expression. The motivation is to rewrite the typing tests of OCaml using this method.

The idea is that one writes:

``` ocaml
let x = 6 * 7
[%expect{|
val x : string
|}]

let x = 1 + 'e';;
[%%expect{||}];;
```

then gives the file to `ocaml -expect-test`, which produces a `.corrected` file containing the real output:

``` ocaml
let x = 42;;
[%%expect{|
val x : int = 42
|}];;

let x = 1 + 'e';;
[%%expect{|
File "", line 1, characters 12-15:
Error: This expression has type char but an expression was expected of type
         int
|}];;
```

To check that the test succeeded, one just has to compare the corrected file with the original one.

The implementation is quite simple and it seems to be enough for typing tests. As an example I converted `typing-immediate` and `typing-poly` to expect tests. I think the result is much more readable than the current .reference files.

For `typing-poly` I tried to have both the output with and without `-principal` in the same ml file, but I'm not sure it's worth it in the end as the file gets a bit harder to read.

### CODE REVIEW

**Comment by gasche** [comment]
I like the results, but I dislike the idea of adding a new (undocumented) feature to the toplevel just for this. Could this be a separate tool or an externally-defined (eg. in the testsuite/ library) directive?

(Also your implementation is hard to read. The two pattern-matchings to extract the extensions are rather boilerplate-y, can't we abstract over these? And I don't understand what (`tag1`, `tag2`) actually are doing by looking at the code).

**Comment by ghost** [comment]
> I like the results, but I dislike the idea of adding a new (undocumented) feature to the toplevel just for this. Could this be a separate tool or an externally-defined (eg. in the testsuite/ library) directive?

Yep, a separate tool is fine. I'll do this

> (Also your implementation is hard to read. The two pattern-matchings to extract the extensions are rather boilerplate-y, can't we abstract over these? And I don't understand what (tag1, tag2) actually are doing by looking at the code).

I need to have another pass, I wanted to see if there was any objections to switching the tests before spending more time on this. [tag1] and [tag2] are the strings tags (as in `{tag|`), for the case with two expectations:

```
<code>;;
[%expect {tag1|
<expectation without -principal>
|tag1}, {tag2|(* principal *)
<expectation with -principal>
|tag2}];;
```

**Comment by ghost** [comment]
I rewrote this as a tool in testsuite/tools

File: testsuite/tools/expect_test.ml [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+(*                                                                     *)
+(*                                OCaml                                *)
+(*                                                                     *)
+(*                 Jeremie Dimino, Jane Street Europe                  *)
+(*                                                                     *)
+(*  Copyright 2016 Jane Street Group LLC                               *)
+(*                                                                     *)
+(*  All rights reserved.  This file is distributed under the terms of  *)
+(*  the GNU Lesser General Public License version 2.1, with the        *)
+(*  special exception on linking described in the file ../LICENSE.     *)
```
Comment (damiendoligez) [inline testsuite/tools/expect_test.ml]: There is a new rule for copyright headers: they must simply mention `LICENSE` without a path.

**Comment by gasche** [comment]
I like the current design better.

The `override_sys_argv` trick is super ugly. Why do you want to have the executable call itself again? I don't understand the specific requirement here, but I would expect a possible solution with either a driver script that calls the tool several times, or maybe some better hooks into the compiler/toplevel's argument parsing machinery (so that you could call the argument-parsing code several time from the same executable, instead of having to re-exec yourself).

**Comment by gasche** [comment]
Also looking quickly at the code it looked like some stuff could be moved to utils/misc (eg. `delete_eol_spaces`). Removing the auxiliary stuff augments the domain-specific density and I find it makes the code easier to read.

File: testsuite/tools/expect_test.ml (hunk 1 above)
Comment (ghost) [inline testsuite/tools/expect_test.ml]: cool, I'm updating the headers

**Comment by ghost** [comment]
OK, I moved `delete_eol_spaces` to `Misc`, as well as `protect_vars` which replaces `Toploop.protect`.

> The override_sys_argv trick is super ugly

Agreed. `Toploop` does the same. Basically you want code executed by the toplevel to see "file.ml" as `Sys.argv.(0)` followed by the arguments passed to the script, not the whole toplevel invocation.

> Why do you want to have the executable call itself again? I don't understand the specific requirement here, but I would expect a possible solution with either a driver script that calls the tool several times, or maybe some better hooks into the compiler/toplevel's argument parsing machinery (so that you could call the argument-parsing code several time from the same executable, instead of having to re-exec yourself).

The file needs to be evaluated once with `!Clflags.principal = true` and once with `Clflags.principal = false`. For simple files it should be enough to evaluate the file twice in the same process and play with `Toploop.toplevel_env`. But if the file contains things like `#load "unix.cma"` then weird things will happen. Using two different processes is a simple way to ensure we start with a clean state. It can be done with a driver script, yes. I just prefer to write OCaml code rather than shell code...

**Comment by gasche** [comment]
Thanks for the clarification. The driver script could be written in OCaml, of course. I think that would be better than the current state..

**Comment by ghost** [comment]
I ended up doing it in the Makefile

**Comment by gasche** [comment]
Thanks, it's nice that you also got a nice win in terms of code simplicity.

Could you add a comment to explain why the `override` hack is there? Maybe we could even expose this hack through the `Toploop` interface, so that you could call it directly instead of duplicating the ugliness.

**Comment by gasche** [comment]
Instead of `Misc.protect_refs [ R (...) ]`, could you use `Misc.R` or `let open Misc`? I'm part of the tribe that believe that warning 40 should be an error by default.

I find this use of GADTs to be a bit over the top; I would have simply sequenced several calls to `protect` together. But well, you did the work.

**Comment by gasche** [comment]
I like the current state of the patch and, barring the relatively minor comments above, I would be fine with merging the PR in trunk -- it does make those testcases sensibly easier to work with. I reviewed the `{opt,}toploop.ml` changes to check they should not introduce regressions.

**Comment by gasche** [comment]
Ah, maybe you could squash some of the review commits (have the Misc additions and factorization first, and then a single commit with the cleaned-up tool code, and then the tests).

**Comment by ghost** [comment]
I exposed `Toploop.override_sys_argv`, removed the `Misc.` prefix in toploop.ml (didn't notice `Misc` was opened in this file) and re-arranged the changes. I didn't change the `Misc.protect_refs` in In expect_test.ml as I'm using type-based disambiguation everywhere in the file

**Comment by garrigue** [comment]
I like very much the way the output is aligned with the input, so that it is easy to understand which test went wrong.
However, do we really need to add this information to the source file?
Wouldn't an output similar to what is done in the manual be sufficient? I.e. having an output file which contains the code too.
My concern is that adding new tests becomes more cumbersome, if you must write the first expected output by hand. Or is everything generated automatically from the start?
Also, I do not like very much having the source file change when the test itself is unchanged.

Not however that I'm a bit biased, because I'm interested in types and error messages, which are often lengthy, and for which the output may change over time. I suppose that if you want to check an expected value it is more natural to write it in the source file.

**Comment by gasche** [comment]
Thanks for chiming in. My understanding is that running the tool leaves a `foo.corrected` file -- test succeeds exactly when it is identical to the input. So you can just write your tests with empty expected outputs, run them, and replace the input file by the `.corrected` file.

**Comment by gasche** [comment]
PS: this was also explained in the [blog post](https://blogs.janestreet.com/testing-with-expectations/) where Yaron presents the version of this tool that is now used at Jane Street. The present PR implements an independent tool, but the workflow is the same.

**Comment by ghost** [comment]
Yeah, the workflow is very much write empty `[%expect]` nodes, then `make && make promote` and check that the nodes contains what you expect.

Having only one file reduces noise in diffs when adding tests or modifying existing ones. This makes reviews easier. Moreover, I think that generally people are more likely to read carefully changes to .ml files than in .expected files.

**Comment by garrigue** [comment]
OK, so the only risk is to forgot one [%expect].
I can live with that.
The problem with principal is more of a concern since we want to keep the tests in sync.

**Comment by garrigue** [comment]
Sorry, I see now that the principal case is handled too. But where do we specify that we want it? It's it always on by default? There may be other options one would want to try.

**Comment by ghost** [comment]
> OK, so the only risk is to forgot one [%expect].

The tool adds a trailing `[%%expect]` if one is missing at the end of the file, so no output is lost

> Sorry, I see now that the principal case is handled too. But where do we specify that we want it? It's it always on by default?

Yeah, it's always on by default. Many tests have both a .reference and a .principal.reference so I though it was simpler to make it the default. In practice it will only be annoying when you only care about one of the cases and the two outputs are different. However, in this case you can just write `#principal <true|false>` at the beginning of the file to force the same setting in both passes

> There may be other options one would want to try.

So most options can  be configured directly from the toplevel via `#`-directives. The only thing that is predefined is that the first pass runs without `-principal` and the second runs with `-principal`. We could allow to configure the set of options for both passes. Either with some Makefile variables, or directly in the .ml file with something like this:

``` ocaml
#first_pass
#warnings "42"
#end

#second_pass
#principal true
#end
```

Personnally I tend to think it's better to keep it simple and have the same thing everywhere, so that we don't have to think about what the two outputs means when reading tests

**Comment by ghost** [comment]
Merged in trunk

**Pull request merged**